### ISSUE TITLE
Aspire 9.0.0 targeting net8.0 forces .NET 9.0.0 library dependencies

### ISSUE THREAD

**Issue by paulomorgado** [dotnet/aspire#6741]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Describe the bug

Aspire 9 Test Project (xUnit) imposes STS libraries on solutions using LTS libraries by taking a dependency on Microsoft.Extensions.Http.Resilience (>= 9.0.0).

That will force a lot of essential dependencies on 9.0.0.

### Expected Behavior

Aspire 9 Test Project (xUnit) does not impose STS libraries on solutions using LTS libraries.

### Steps To Reproduce

Create an Aspire 9 Test Project (xUnit) targeting net8.0 on a solution with all projects targeting net8.0 (or bellow) with [central package management](https://learn.microsoft.com/en-us/nuget/consume-packages/Central-Package-Management) with [version override](https://learn.microsoft.com/en-us/nuget/consume-packages/Central-Package-Management#overriding-package-versions) disabled.

### .NET Version info

```
.NET SDK:
 Version:           8.0.404
 Commit:            7b190310f2
 Workload version:  8.0.400-manifests.4d7d414a
 MSBuild version:   17.11.9+a69bbaaf5

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19045
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\8.0.404\

.NET workloads installed:
Configured to use loose manifests when installing new manifests.
 [aspire]
   Installation Source: SDK 8.0.400, VS 17.13.35507.96, VS 17.12.35506.116
   Manifest Version:    8.2.2/8.0.100
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100\microsoft.net.sdk.aspire\8.2.2\WorkloadManifest.json
   Install Type:        FileBased


Host:
  Version:      9.0.0
  Architecture: x64
  Commit:       9d5a6a9aa4

.NET SDKs installed:
  6.0.428 [C:\Program Files\dotnet\sdk]
  8.0.404 [C:\Program Files\dotnet\sdk]
  9.0.100 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 8.0.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 9.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 8.0.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 9.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 8.0.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 9.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  C:\src\services-platform\connext\connector\global.json

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```

**Comment by rudiv** [comment]
The majority of the 9.0.0 libraries still have a net8.0 target (including Microsoft.Extensions.Http.Resilience referenced)?

**Comment by paulomorgado** [comment]
> The majority of the 9.0.0 libraries still have a net8.0 target (including Microsoft.Extensions.Http.Resilience referenced)?

With an STS support policy instead of an LTS support policy.

**Comment by rudiv** [comment]
Right, but you're aware of the [Aspire support policy](https://dotnet.microsoft.com/en-us/platform/support/policy/aspire) right? Which is that it's not supported as soon as a new version is released?

**Comment by paulomorgado** [comment]
> Right, but you're aware of the [Aspire support policy](https://dotnet.microsoft.com/en-us/platform/support/policy/aspire)? Which is that it's not supported as soon as a new version is released?

I am very aware of Microsoft support policies.

And Aspire is forcing STS libraries on a solution that must use only LTS libraries.

**Comment by rolfik-mycronic** [comment]
We are in the same situation.
We .NET 8 solution due to LTS, but would benefit from Aspire 9.

**Comment by paulomorgado** [comment]
Any update on this, @joperezr?

**Comment by joperezr** [comment]
Sorry for not replying here sooner, @paulomorgado. This is very much top of mind for us, and @terrajobst and others have been looking at what the best option is for this going forward. I don't think we are there yet, but I hope we can have an answer for this soon.

**Comment by paulomorgado** [comment]
Sorry to nag, @joperezr and @terrajobst, have you come to a conclusion on this?

**Comment by joperezr** [comment]
No updates yet, but we continue to work on this, we have been reviewing some proposals. In some of our discussions, it would be actually helpful to understand what is the biggest concern here. While I understand some of the implications that referring to 9.0 packages is, I think it would help validate our assumptions if we understand your position better. Is the primary issue support? If so, could you elaborate? Is it instead having to adapt to breaking changes? Any and all details you provide will help in the design for our proposal.

**Comment by paulomorgado** [comment]
> No updates yet, but we continue to work on this, we have been reviewing some proposals. In some of our discussions, it would be actually helpful to understand what is the biggest concern here. While I understand some of the implications that referring to 9.0 packages is, I think it would help validate our assumptions if we understand your position better. Is the primary issue support? If so, could you elaborate? Is it instead having to adapt to breaking changes? Any and all details you provide will help in the design for our proposal.

@joperezr, the existence of LTS and STS on itself is reason enough to respect it and the commitments made to users.

A simple dependency can bring a cascade of other dependencies. It's not about one single dependency.

A company can have a mandate of using only LTS dependencies and choose to use Central Package Management without version override permissions, which will exclude the use of Aspire 9.*.

There are [documented breaking changes moving from .NET 8 to .NET9](https://learn.microsoft.com/en-us/dotnet/core/compatibility/9.0) and most probably there will be breaking changes moving from .NET9 to .NET10.

Given that support of .NET8 ends on November 10 2026 and support of .NET9 ends on May 12 2026, forcing users on .NET9 libraries when they choose to be on LTS (.NET8) is forcing them to 2 breaking changes or 1 breaking change and being unsupported for 6 months. The latter is not likely to happen as the dependencies will most certainly change to .NET10.

And all this for no practical benefit.

(I'm having flashbacks of ".NET Framework 4.6.1 is .NET Standard 2.0" and "HttpClient on .NET Framwork" flashbacks that I never expected to have. :smile:)

**Comment by eerhardt** [comment]
This is a little confusing IMO, but Microsoft.Extensions.Http.Resilience ships out of the dotnet/extensions repo and isn't in a LTS/STS support policy. Instead it is on a "latest" or "tip" only support.

https://dotnet.microsoft.com/en-us/platform/support/policy/extensions

> At any time, only the latest release of .NET Platform Extensions is supported by Microsoft. For example, support for .NET Platform Extensions 8.0 ended when .NET Platform Extensions 8.1 was released, and support for .NET Platform Extensions 8.10 ends when .NET Platform Extensions 9.0 is released.

So the `9.1.0` is the currently supported version as of this writing. Note that it doesn't bring in the 9.0.0 libraries of things in a "shared framework". Instead, it has a different set of dependencies based on the TFM being targetted:

![Image](https://github.com/user-attachments/assets/d633202a-1305-48fb-a44c-279e448497e5)

So when you are on `net8.0`, you will still get the LTS versions of the libraries that are contained in the runtime and ASP.NET core shared frameworks.

Basically, the ".NET Platform Extensions" libraries version and are supported the same way as .NET Aspire - the latest version is supported, and versions always support the current supported LTS version.

**Comment by paulomorgado** [comment]
Thanks, @eerhardt.

I realize now that I didn't investigate this thoroughly. My apologies for that.

I can see now that proper care was taken on the dependencies from .NET:

![Image](https://github.com/user-attachments/assets/c834cb7a-ae34-4138-b871-7e5780b48fea)

Since we are on the subject, I'd like your opinion on this: https://github.com/open-telemetry/opentelemetry-dotnet/issues/5973

**Comment by davidfowl** [comment]
@DamianEdwards this is yours

**Comment by paulomorgado** [comment]
My confusions was that these packages are covered by [Microsoft support for .NET Platform Extensions](https://dotnet.microsoft.com/en-us/platform/support/policy/extensions), not [.NET Support Policy](https://dotnet.microsoft.com/en-us/platform/support/policy).

Just make sure the [Microsoft support for .NET Platform Extensions](https://dotnet.microsoft.com/en-us/platform/support/policy/extensions) list is up to date.
